RefsViews/Branches/BranchesModel.cpp

```diff
@@ -165,68 +171,110 @@ bool BranchesModel::hasChildren( const QModelIndex& parent ) const
 }
 
+void BranchesModel::insertRef(bool notify, const Git::Reference &ref)
+{
+    RefScope* scope = scopeForRef( ref );
+    Q_ASSERT( scope );
+
+    QStringList parts = ref.shorthand().split( QChar( L'/' ) );
+    if ( parts.count() == 1 )
+    {
+        insertBranch( notify, scope, parts.last(), ref );
+        return;
+    }
+
+    RefItem* ns = scope;
+    for( int j = 0; j < parts.count() - 1; j++ )
+    {
+        RefItem* next = NULL;
+        QString partName = parts[ j ];
+        foreach( RefItem* nsChild, ns->children )
+        {
+            if( nsChild->text() == partName )
+            {
+                next = nsChild;
+                break;
+            }
+        }
+
+        if( !next )
+        {
+            next = insertNamespace( notify, ns, partName );
+        }
+        ns = next;
+    }
+
+    Q_ASSERT( ns );
+
+    insertBranch( notify, ns, parts.last(), ref );
+}
+
+RefItem* BranchesModel::insertNamespace(const bool notify, RefItem* parent, const QString& name)
+{
+    RefItem* next = NULL;
+    if ( notify ) {
+        int fr = parent->children.count();
+        beginInsertRows( index( parent ), fr, fr );
+    }
+
+    next = new RefNameSpace( parent, name );
+
+    if ( notify ) {
+        endInsertRows();
+    }
+    return next;
+}
+
+void BranchesModel::insertBranch(const bool notify, RefItem* parent, const QString& name, const Git::Reference& ref)
+{
+    if ( notify ) {
+        int row = parent->children.count();
+        beginInsertRows( index( parent ), row, row );
+    }
+
+    new RefBranch( parent, ref );
+
+    if (notify) {
+        endInsertRows();
+    }
+}
+
+RefScope* BranchesModel::scopeForRef(const Git::Reference& ref) const
+{
+    RefItem* scope = NULL;
+    if ( ref.isLocal() )        scope = mRoot->children[0];
+    else if ( ref.isRemote() )  scope = mRoot->children[1];
+    else scope = mRoot->children[2];
+
+    return static_cast< RefScope* >( scope );
+}
+
 void BranchesModel::rereadBranches()
 {
     beginResetModel();
 
-    Git::Repository repo = mData->repository();
-
     qDeleteAll( mRoot->children );
-    Q_ASSERT( mRoot->children.isEmpty() );
+    mRoot->children.clear();
+
+    new RefScope( mRoot, tr( "Local" ) );
+    new RefScope( mRoot, tr( "Remote" ) );
+    new RefScope( mRoot, tr( "Tags" ) );
 
-    if( repo.isValid() )
+    RM::Repo* repo = mData->repository();
+
+    // TODO: migrate to RM::Repo
+    Git::Repository gitRepo = repo ? repo->gitRepo() : Git::Repository();
+
+    if( gitRepo.isValid() )
     {
         Git::Result r;
-        Git::ReferenceList sl = repo.allReferences( r );
+        Git::ReferenceList sl = gitRepo.allReferences( r );
         if( !sl.isEmpty() )
         {
-            RefScope* scopeLocal = new RefScope( mRoot, tr( "Local" ) );
-            RefScope* scopeRemote = new RefScope( mRoot, tr( "Remote" ) );
-            RefScope* scopeOther = new RefScope( mRoot, tr( "Tags" ) );
-
             for( int i = 0; i < sl.count(); ++i )
```

Comment:
Actually my comment was total bullshit :). The `refCreated` events are not fired on opening/reading the repo, because `repoEventsBlocked` prevents that for good reason.

Well ... what do you mean by "hook into levels of RepoManager". Isn't this the same thing in green?